core/templates/components/state-directives/hint-editor/hint-editor.component.ts

```diff
@@ -16,97 +16,111 @@
+import { Hint } from 'domain/exploration/HintObjectFactory';
 
-angular.module('oppia').component('hintEditor', {
-  bindings: {
-    hint: '=',
-    getIndexPlusOne: '&indexPlusOne',
-    getOnSaveFn: '&onSave',
-    showMarkAllAudioAsNeedingUpdateModalIfRequired: '='
-  },
-  template: require(
-    'components/state-directives/hint-editor/hint-editor.component.html'),
-  controllerAs: '$ctrl',
-  controller: [
-    'ContextService', 'EditabilityService',
-    'ExternalSaveService', 'StateHintsService',
-    function(
-        ContextService, EditabilityService,
-        ExternalSaveService, StateHintsService) {
-      var ctrl = this;
-      ctrl.directiveSubscriptions = new Subscription();
-      ctrl.openHintEditor = function() {
-        if (ctrl.isEditable) {
-          ctrl.hintMemento = angular.copy(ctrl.hint);
-          ctrl.hintEditorIsOpen = true;
-        }
-      };
-
-      ctrl.isHintLengthExceeded = function() {
-        // TODO(#13764): Edit this check after appropriate limits are found.
-        return (ctrl.hint.hintContent._html.length > 10000);
-      };
-
-      ctrl.saveThisHint = function() {
-        ctrl.hintEditorIsOpen = false;
-        var contentHasChanged = (
-          ctrl.hintMemento.hintContent.html !==
-          ctrl.hint.hintContent.html);
-        ctrl.hintMemento = null;
-        if (contentHasChanged) {
-          var hintContentId = ctrl.hint.hintContent.contentId;
-          ctrl.showMarkAllAudioAsNeedingUpdateModalIfRequired(
-            [hintContentId]);
-        }
-        ctrl.getOnSaveFn()();
-      };
-
-      ctrl.cancelThisHintEdit = function() {
-        ctrl.hint.hintContent =
-          angular.copy(ctrl.hintMemento.hintContent);
-        ctrl.hintMemento = null;
-        ctrl.hintEditorIsOpen = false;
-      };
-
-      ctrl.$onInit = function() {
-        ctrl.directiveSubscriptions.add(
-          ExternalSaveService.onExternalSave.subscribe(() => {
-            if (ctrl.hintEditorIsOpen &&
-                  ctrl.editHintForm.$valid) {
-              ctrl.saveThisHint();
-            }
-          }));
-        ctrl.isEditable = EditabilityService.isEditable();
-        ctrl.StateHintsService = StateHintsService;
-        ctrl.editHintForm = {};
-        ctrl.hintEditorIsOpen = false;
-
-        ctrl.HINT_FORM_SCHEMA = {
-          type: 'html',
-          ui_config: {
-            hide_complex_extensions: (
-              ContextService.getEntityType() === 'question')
-          }
-        };
-
-        ctrl.hintMemento = null;
-      };
-      ctrl.$onDestroy = function() {
-        ctrl.directiveSubscriptions.unsubscribe();
-      };
+interface HintFormSchema {
+  type: string;
+  'ui_config': object;
+}
+
+@Component({
+  selector: 'oppia-hint-editor',
+  templateUrl: './hint-editor.component.html'
+})
+export class HintEditorComponent implements OnInit, OnDestroy {
+  @Input() hint: Hint;
+  @Input() indexPlusOne!: number;
+  @Input() saveHint;
```

Comment:
Please add type here.

Reply:
these are functions replaced with output now as you explained `We try to remove the pattern of passing down functions from parent components to child. Instead, use @output here and listen to that event in the parent component and call the original function in the parent component itself.
Here and elsewhere.`